### ISSUE TITLE
Artificially trigger and capture sites' 404 pages

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/legacy.httparchive.org#123]
To help answer the question "How many sites have a custom 404 page?" and the like, it may be useful to intentionally trigger a 404 page by hitting a known non-existent page during the test just to see what happens.

404 pages tend to fail fast so I wouldn't expect the overall crawl duration to be significantly affected.

@pmeenan would this be as straightforward as using scripting to do something like this?

```
navigate http://www.example.com/
navigate http://www.example.com/http-archive-404-test.html
```

My only concern would be leaving the page before we're done capturing what would normally be captured. Maybe we could instead do it with an `execAndWait` command to fetch the 404 page, which would also need to be timed to not interfere with the normal page load.

**Comment by igrigorik** [comment]
Wouldn't that increase overhead by ~25%? We'd be going from 3 to 4 pages.. which is non-trivial, plus logistics around how responses are stored, etc?

**Comment by rviscomi** [comment]
I was thinking of it as just adding another request to the page load. Since the average page has ~100 requests it should be more like an additional 1%.

**Comment by pmeenan** [comment]
Some would arbitrarily redirect as well so it doesn't feel all that easily detectable.  It's also not just an additional request if you are doing a new navigation - it's a new page load which may or may not 404.

I'd love to see a really strong use case for the data before considering it given the dev effort that would also have to go into to do the measurement.

**Comment by rviscomi** [comment]
Good points. The use case (mentioned off thread) is SEO analysis. Google's [SEO Starter Guide](https://static.googleusercontent.com/media/www.google.com/en//webmasters/docs/search-engine-optimization-starter-guide.pdf) mentions having a useful 404 page and that seems like a blind spot in HA's data.

Lighthouse actually has a [FR](https://github.com/GoogleChrome/lighthouse/issues/2091) to add SEO audits, so this issue would become obsolete if that were to happen.

As for a potential implementation if needed, one way around adding a new page load / navigation could be a custom metric script or similar that runs `fetch('/http-archive-404-test.html')` ignoring the response. That way, the request should still go through WPT and HA should also have the response body.

Not super attached to the idea either. Would just be nice to have if it's easy to implement and doesn't adversely affect crawl/storage. But in general, it'd be great if LH took it off our hands :D

**Comment by pmeenan** [comment]
Custom metrics need to be completely blocking and can't do async operations (like a fetch).

Honestly, scanning for a 404 page would probably be a lot cleaner to just do directly in the python code (no need to ACTUALLY use Chrome - just send the right UA string.  It could presumably even do it in parallel with processing of the test result data after the main page has loaded (probably just a few hours of work to implement and test).  Would still need to have some place to put the resulting data though and what you were trying to gather from it.

**Comment by Themanwithoutaplan** [comment]
404 is a status code. Whether or not there is an error page attached is secondary: I fondly remember WebPositive Haiku error message that the page didn't exist. It tells users the resource isn't there and website owners what people are looking for (404s sometimes arise when pages are deleted or renamed but more often when spiders are looking for hidden pages or potential exploits). Trying to the trigger them with specially crafted URLs is, in my opinion, not playing nice and way outside HA's remit.

If you're really interested in this then, as Pat says, running a low level script to collect URLs would be the way to go and scales very well.
